src/vector.h

```diff
@@ -8,13 +8,13 @@ template <typename T> class Vector {
 public:
   using Value = T;
 
-  Vector() = default;
-  Vector(std::int32_t n) { data.resize(n); }
+  Vector() : data{} {}
+  Vector(std::int32_t n) : data(static_cast<std::size_t>(n)) {}
```

Comment:
```c++
Vector<int> v(-1);
```

```bash
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
Aborted (core dumped)
```
Just caught my eye ^. Does signed int really needed? 
Why do we need this `std::vector` wrapper at all ?

Reply:
I think we started out with a matrix wrapper https://github.com/mapbox/node-or-tools/blob/4364bd69e35c5e25504f0d6dd2c9d1ac829f5a57/src/matrix.h and then just went with it.

Re. your question for signed ints: this project uses signed ints everywhere and we consider unsigned ints in APIs like vector or for anything else then bitmasks / bytes a design mistake.

Your `vector v(-1)` is a good example for this: the signed `-1` integer literal gets converted to `numeric_limits<size_t>::max()` and then the vector constructor has no way of asserting `n >= 0` for the `vector(n)` constructor. Similarly in other situations with unsigned ints the sanitizers can't catch overflows or other issues easily.

It's unfortunate that the stdlib is unsigned in general but it's a mistake we don't have to repeat.

Reply:
Ok, thanks for the background.
> It's unfortunate that the stdlib is unsigned in general but it's a mistake we don't have to repeat.
Do you have some links etc ? I'd love to catch up on this as it comes up quite a lot ^

Reply:
I _think_ there were some talks and discussions at C++ conferences about this; some from the Google folks in the context of their sanitizers and detecting these issues. And this discussion is going on in the larger C++ community (irc, mailinglists, twitter, reddit, ...) but I don't have good resources to point you at of the top of my head, sorry.

Reply:
@daniel-j-h So, part of the cleanup require here was fixing a bunch of signed/unsigned cast warnings because we've added `-Wconversion` to the build flags.  There were a *tonne* of places where signed/unsigned values were being converted, a couple very dubiously.

I've fixed all the warnings, but to me, the API feels a bit off - there are very few places where we actually accept negative numbers - why would we make these parameters signed at all?

Reply:
See conversation and arguments above. Unsigneds silently wrap around and you can't detect errors with them. Unsigneds should only be used for bitmasks / bytes.

Reply:
@danpat unsigned wraps around - that's exactly the problem, you can't detect an issue when it overflows. Signed overflow is undefined which allows the sanitizers (or other checkers) to catch it. You usually never want the overflow semantics; a `vector(-1)` should error out and not try to allocate `numeric_limits<size_t>::max()` memory.